Actioning feedback from Survey on iFrames (#1855)
act-rules/act-rules.github.io: _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md

---
id: akn7bn

> name: Iframe with interactive elements are in tab-order

Reviewer: ```suggestion
name: Interactive elements within iframe are in tab-order
```

@Jym77, per HTML specs and your previous feedback:

> iframe are not necessarily [focusable areas](https://html.spec.whatwg.org/multipage/interaction.html#focusable-area) (this is actually UA dependent due to the the "or the element is determined by the user agent to be focusable" bit 😞), and therefore are (possibly) dropped out of the flattened scope (and the final tab-order) of the full page.

from title seems that we leave browsers with the responsibility to decide if an iframe is passing or failing this rule.
In fact, if you create an iframe with a simple input as content (e.g. `<input type="text" name="test" aria-label="test">`),opening it in FireFox the iframe is tabbable (and consistent with this rule title), if instead you open it in Chrome the iframe is not tabbable (and seems not consistent with this rule title).

The final goal is to allow users interacting with the iframe content, so I would suggest to use "interactive elements" as subject of this rule.

Reviewer: On the other hand, this leads to a bit of discrepancy between rule title and "test target"/applicability 🤔

> name: Iframe with interactive elements are not excluded from tab-order

Reviewer 2: ```suggestion
name: Iframe with interactive elements is not excluded from tab-order
```

Reviewer 2: ```suggestion
name: Iframe with interactive elements does not have a negative tabindex
```

rule_type: atomic
description: |

>   This rule checks that `iframe` elements which contain an interactive (tabbable) element are themselves tabbable.

Reviewer: ```suggestion
  This rule checks that interactive elements contained in `iframe` elements are tabbable.
```

same as above

## Applicability

> This rule applies to any `iframe` element that [contains](#akn7bn:contain) elements that are both [visible][] and part of the [sequential focus navigation][].

Reviewer 3: Expliciting the logic.
```suggestion
This rule applies to any `iframe` element that [contains](#akn7bn:contain) at least one element for which all the following are true:
- the element is [visible][]; and
- the element is part of the [flattened tabindex-ordered focus navigation scope][] of the `iframe`.
```

> An element is <dfn id="akn7bn:contain">contained</dfn> in a [nested browsing context][] if its [owner document][] is the [container document][] of the [nested browsing context][]. The elements (inside the iframe) are part of the [flattened tabindex-ordered focus navigation scope].

Reviewer 3: ```suggestion
An element is <dfn id="akn7bn:contain">contained</dfn> in a [nested browsing context][] if its [owner document][] is the [container document][] of the [nested browsing context][]. 

The [flattened tabindex-ordered focus navigation scope][] of an `iframe` is essentially the [sequential focus navigation][] order inside its content.
```

## Expectation

> For each test target, the [nested browsing context][] does not have a negative number as a `tabindex` [attribute value][].

Reviewer 3: Agree with @giacomo-petri The negative tabindex is set on the `iframe` element, aka the browsing context container, not on the nested browsing context (which is an abstract notion not really "existing" in the HTML code and not having attributes).

```suggestion
None of the test target has a negative number as a `tabindex` [attribute value][].
```

## Applicability

This rule applies to any `iframe` element that [contains](#akn7bn:contain) at least one element for which all the following are true:
- the element is [visible][]; and
- the element is part of the [flattened tabindex-ordered focus navigation scope][] of the `iframe`.

An element is <dfn id="akn7bn:contain">contained</dfn> in a [nested browsing context][] if its [owner document][] is the [container document][] of the [nested browsing context][]. 

The [flattened tabindex-ordered focus navigation scope][] of an `iframe` is essentially the [sequential focus navigation][] order inside its content.

## Expectation

> None of the test target has a negative number as a `tabindex` [attribute value][].

Reviewer 3: I'm actually considering whether we should rather request the iframe to be part of the tab-order of the top-level document. That would be a bit more generic for ways other than negative tabindex  to remove stuff from tab-order. But it also seems that currently negative tabindex (on the iframe) is the only way to do so (for tabbable inside an iframe). So that would complicate the rule with technical mumbo-jumbo without really improving it 🤔

Author: I think this might be something we can discuss in the next meeting?

Reviewer 4: +1 to keeping as-is, if we are going to complicate things unnecessarily.

## Applicability

This rule applies to any `iframe` element that [contains](#akn7bn:contain) at least one element for which all the following are true:

- the element is [visible][]; and
- the element is part of the [sequential focus navigation order][] of the `iframe`'s [document][].

An element is <dfn id="akn7bn:contain">contained</dfn> in a [nested browsing context][] if its [owner document][] is the [container document][] of the [nested browsing context][].

## Expectation

> The test target does not have a negative number as a `tabindex` [attribute value][].

Reviewer 5: LGTM, although I want to note that `inert` might also now be able to trigger this situation. I opened a separate issue for it: #1965

## Applicability

This rule applies to any `iframe` element that [contains](#akn7bn:contain) at least one element for which all the following are true:

- the element is [visible][]; and

> - the element is part of the [flattened tabindex-ordered focus navigation scope][] of the `iframe`.

Reviewer 6: Do we need to keep the following sentence about nested browsing context?

Reviewer 3: I think we do.
There are some technical subtelties in what owns what and what is inside what with browsing context containers, browsing contexts, and documents. The clarification makes it explicit what is meant here.
I also feel that putting it as a "next line clarification" does not really block readability and is only providing details for the technically savvy readers.

> - the element is part of the [flattened tabindex-ordered focus navigation scope][] of the `iframe`.

Reviewer 5: I think this got unnecessarily complicated. 

```suggestion
- the element is part of the [sequential focus navigation][] of the `iframe`.
```

An element is <dfn id="akn7bn:contain">contained</dfn> in a [nested browsing context][] if its [owner document][] is the [container document][] of the [nested browsing context][].

> The [flattened tabindex-ordered focus navigation scope][] of an `iframe` is essentially the [sequential focus navigation][] order inside its content.

Reviewer 5: You don't need this. All documents have sequential focus navigation, an iframe has a document, ergo. Iframes can have sequential focus navigation.

Reviewer 3: Indeed 🙈 
I somehow got lost in technicalities and forgot that the sequential focus navigation is indeed exactly the FTOFNS 🙃 
(well, not exactly, because the former is an order and the later is a list 🤓, but close enough)

## Background

> Inside a page, each nested document (mostly, content of an `iframe` element) has its own [sequential focus navigation order][] which is the local tab-order inside this document. Then, all of these are grouped together in a [flattened tabindex-ordered focus navigation scope][] which is effectively the tab-order of the full page. Setting the `tabindex` attribute of an `iframe` element to a negative value effectively excludes its content from the global tab-order of the page (when building the [tabindex-ordered focus navigation scope][] of the page). That is, a `button` may be in tab-order of a document, but if that document is included in another page via an `iframe` with negative `tabindex`, then the `button` is not in the tab-order of that page; and therefore not keyboard accessible.

Reviewer 5: I think this is still a bit tricky. I tried to rephrase this, to hopefully improve the readability.

```suggestion
Setting the `tabindex` attribute of an `iframe` element to a negative value effectively excludes the content of that `iframe` from the tab-order of the page. A `button` may be in the tab-order of inside an `iframe`, but if the `iframe` is taken from the tab-order, the `button` is effectively keyboard inaccessible. 

Each document, including document inside an `iframe` has its own [sequential focus navigation order][]. These focus orders are combined to get the page's global tab-order (called the [flattened tabindex-ordered focus navigation scope][]). For an `iframe` with a negative tabindex, its sequential focus navigation order is not included in the page's global tab-order.
```

## Background

> Inside a page, each nested document (mostly, content of an `iframe` element) has its own [tabindex-ordered focus navigation scope][] which is the local tab-order inside this document. Then, all of these are grouped together in a [flattened tabindex-ordered focus navigation scope][] which is effectively the tab-order of the full page. By setting the `tabindex` attribute of a [focus navigation scope owner][] (mostly, an `iframe` element) to a negative value, it is excluded from its containing document's [tabindex-ordered focus navigation scope][]; and therefore its own [tabindex-ordered focus navigation scope][] is excluded from the global [flattened tabindex-ordered focus navigation scope][]. That is, a `button` may be in tab-order of a document, but if that document is included in another page via an `iframe` with negative `tabindex`, then the `button` is not in the tab-order of that page; and therefore not keyboard accessible.

Reviewer 5: This text is far more complicated than it needs to be to explain the problem; Tabindex=-1 on iframes means things in the iframe cannot be focused with keyboard navigation.

Author: To be discussed in the ACT-R meeting

Reviewer 3: I've simplified this note.
I kept a bit of the technicalities and links because this corresponds to discussions we had during [CG call](https://www.w3.org/2022/07/28-act-r-minutes.html#item05) and [in this PR](https://github.com/act-rules/act-rules.github.io/pull/1855#issuecomment-1199064988), so I feel we need to keep trace of it.

#### Passed Example 1

> This `iframe` element contains a link that is not part of [sequential focus navigation][] because of its `tabindex`.

Reviewer: Should we mark also this example as inapplicable now, due to the new definition "This rule applies to any `iframe` element that [contains](#akn7bn:contain) elements that are both [visible][] and part of the [sequential focus navigation][]."?

We should add instead an example with an iframe with no tabindex attribute (or tabindex="0", or both) and both visible and SFN  content.

#### Passed Example 1

This `iframe` element contains a link that is not part of [sequential focus navigation][] because of its `tabindex`.

```html

> <iframe tabindex="-1" srcdoc="<a href='/' tabindex='-1'>Home</a>"></iframe>

Reviewer 3: I agree that with the change, this is now Inapplicable and should be moved accordingly.

```

### Failed

#### Failed Example 1

> This `iframe` element has a link that is part of [sequential focus navigation][].

Reviewer 3: Maybe we can boost the description with the technical mumbo-jumbo…
```suggestion
This `iframe` element contains a [visible][] link that is part of its [flattened tabindex-ordered focus navigation scope][]. Note that the link is not part of the document's [sequential focus navigation][] because the containing `iframe` is excluded from the document's [tabindex-ordered focus navigation scope](https://html.spec.whatwg.org/multipage/interaction.html#tabindex-ordered-focus-navigation-scope) due to its negative `tabindex`.
```

```html
<iframe tabindex="-1" srcdoc="<a href='/'>Home</a>"></iframe>
```

### Inapplicable

#### Inapplicable Example 1

This `iframe` element does not have a `tabindex` [attribute value][] that is a negative number

```html

> <iframe tabindex="0" srcdoc="<a href='/'>Home</a>"></iframe>

Reviewer 3: This one has no become a Passed example since the iframe has visible and tabbable content.

#### Passed Example 1

This `iframe` element does not have a `tabindex` [attribute value][] that is a negative number

```html
<iframe tabindex="0" srcdoc="<a href='/'>Home</a>"></iframe>
```

### Failed

#### Failed Example 1

> This `iframe` element contains a [visible][] link that is part of its [flattened tabindex-ordered focus navigation scope][]. Note that the link is not part of the document's [sequential focus navigation][] because the containing `iframe` is excluded from the document's [tabindex-ordered focus navigation scope](https://html.spec.whatwg.org/multipage/interaction.html#tabindex-ordered-focus-navigation-scope) due to its negative `tabindex`.

Reviewer 3: I'm also considering whether that inlined note should be made into a full Background note (slightly changed), since it explain the full intend of the rule: stuff that is in the FTOFNS of an iframe should be in the tab-order of the web page.

Author: I think this might be something we can discuss in the next meeting?

Reviewer 4: +1 to adding this as a background note

```html
<iframe tabindex="-1" srcdoc="<a href='/'>Home</a>"></iframe>
```

### Inapplicable

#### Inapplicable Example 1

> This `iframe` element contains no content that is part of [sequential focus navigation][].

Reviewer 3: ```suggestion
This `iframe` element contains no content that is part of its [flattened tabindex-ordered focus navigation scope][].
```
(in Failed Example 1, the link is part of the FTOFNS (!) but not of the sequential focus order, due to the negative tabindex of the iframe; the difference is that here there is nothing in th FTOFNS of the iframe, so we don't care about the negative tabindex).

```html
<iframe tabindex="-1" srcdoc="<h1>Hello world</h1>"></iframe>
```

#### Inapplicable Example 2

This `iframe` element contains no [visible][] content because the iframe is hidden.

```html
<iframe tabindex="-1" hidden srcdoc="<a href='/'>Home</a>"></iframe>
```

#### Inapplicable Example 3

This `iframe` element contains no [visible][] content because of the small size of the iframe.

```html
<iframe tabindex="-1" width="1" height="1" srcdoc="<a href='/'>Home</a>"></iframe>
```

#### Inapplicable Example 4

> This `iframe` element contains a link that is not part of [sequential focus navigation][] because of its `tabindex`.

Reviewer 3: ```suggestion
This `iframe` element contains a link that is not part of its [flattened tabindex-ordered focus navigation scope][] because of its `tabindex`.
```

<iframe tabindex="0" srcdoc="<a href='/'>Home</a>"></iframe>
```

#### Inapplicable Example 2

> This `iframe` element contains no content that is part of [sequential focus navigation][].

Reviewer 3: The Applicability does not mention "content that is part of sequential focus navigation".
So this example is currently Applicable as it is an `iframe` with negative tabindex, and it passes the Expectation as there is no focusable content in it.

Author: I have switched the expectation and applicability around to help make these valid.

```html
<iframe tabindex="-1" srcdoc="<h1>Hello world</h1>"></iframe>
```

#### Inapplicable Example 3

> This `iframe` element contains no [visible][] content because the iframe is hidden.

Reviewer 3: The applicability does not mention the `iframe` to be visible or not.

Author: I have switched the expectation and applicability around to help make these valid.
